Change slice types to interface when not the same? - panic: reflect.AppendSlice: interface {} != string

Greetings,

Do you think it would be alright to add some more code to vm.go to check if slices/array as the same type before appending them? Change them to interface automatically when they are not?

If it is alright to do the automatic conversion to interface, I can see about working on a PR.

```Go
package main

import (
	"fmt"
	"reflect"

	"github.com/mattn/anko/builtins"
	"github.com/mattn/anko/vm"
)

func main() {
	var value reflect.Value

	vmEnv := vm.NewEnv()
	core.LoadAllBuiltins(vmEnv)

	input := []string{"one", "two"}

	err := vmEnv.Define("input", input)
	if err != nil {
		fmt.Printf("Define error: %v", err)
		return
	}

	script := `
output = []

output += input
`

	_, err = vmEnv.Execute(script)
	if err != nil {
		fmt.Printf("Get error: %v", err)
		return
	}

	value, err = vmEnv.Get("output")
	if err != nil {
		fmt.Printf("Get error: %v", err)
		return
	}

	fmt.Println(value)
}

```

```
panic: reflect.AppendSlice: interface {} != string

goroutine 1 [running]:
reflect.typesMustMatch(0x8153b4, 0x13, 0x9b7160, 0x7a4d20, 0x9b7160, 0x78d720)
	C:/app/Go/src/reflect/value.go:1790 +0x19c
reflect.AppendSlice(0x7876a0, 0xc0420a8e60, 0x97, 0x788860, 0xc0420a8d40, 0x97, 0x0, 0x0, 0x0)
	C:/app/Go/src/reflect/value.go:1844 +0x123
vendor/github.com/mattn/anko/vm.invokeExpr(0x9b18a0, 0xc04200c400, 0xc042010500, 0x0, 0x0, 0x0, 0x0, 0x0)
	C:/Users/username/Documents/Eclipse/src/vendor/github.com/mattn/anko/vm/vm.go:1207 +0x230d
vendor/github.com/mattn/anko/vm.invokeExpr(0x9b1860, 0xc04200c3c0, 0xc042010500, 0x0, 0x0, 0x0, 0x0, 0x0)
	C:/Users/username/Documents/Eclipse/src/vendor/github.com/mattn/anko/vm/vm.go:1121 +0x6b97
vendor/github.com/mattn/anko/vm.RunSingleStmt(0x9b1b20, 0xc0420a8e20, 0xc042010500, 0x0, 0x0, 0x0, 0x0, 0x0)
	C:/Users/username/Documents/Eclipse/src/vendor/github.com/mattn/anko/vm/vm.go:130 +0x476a
vendor/github.com/mattn/anko/vm.Run(0xc0420a8e40, 0x2, 0x2, 0xc042010500, 0x0, 0x0, 0x0, 0x0, 0x0)
	C:/Users/username/Documents/Eclipse/src/vendor/github.com/mattn/anko/vm/vm.go:95 +0x286
vendor/github.com/mattn/anko/vm.(*Env).Execute(0xc042010500, 0x81a160, 0x1e, 0x0, 0x0, 0x0, 0x0, 0x0)
	C:/Users/username/Documents/Eclipse/src/vendor/github.com/mattn/anko/vm/env.go:257 +0x14f
main.main()
	C:/Users/username/Documents/Eclipse/src/test/main.go:32 +0x2ce
```

vm.go code in question:
```Go
			if (lhsV.Kind() == reflect.Array || lhsV.Kind() == reflect.Slice) && (rhsV.Kind() != reflect.Array && rhsV.Kind() != reflect.Slice) {
				return reflect.Append(lhsV, rhsV), nil
			}
			if (lhsV.Kind() == reflect.Array || lhsV.Kind() == reflect.Slice) && (rhsV.Kind() == reflect.Array || rhsV.Kind() == reflect.Slice) {
				return reflect.AppendSlice(lhsV, rhsV), nil
			}
```

Best regards,

Michael

--- a/vm/vmExpr.go
+++ b/vm/vmExpr.go
@@ -605,7 +605,7 @@ func invokeExpr(expr ast.Expr, env *Env) (reflect.Value, error) {
 				return reflect.ValueOf(toString(lhsV) + toString(rhsV)), nil
 			}
 			if (lhsV.Kind() == reflect.Array || lhsV.Kind() == reflect.Slice) && (rhsV.Kind() != reflect.Array && rhsV.Kind() != reflect.Slice) {
-				rhsT := rhsV.Type().Elem()
+				rhsT := rhsV.Type()
 				lhsT := lhsV.Type().Elem()
 				if lhsT.Kind() != rhsT.Kind() {
 					if !rhsT.ConvertibleTo(lhsT) {

--- a/vm/vmExpr.go
+++ b/vm/vmExpr.go
@@ -605,9 +605,28 @@ func invokeExpr(expr ast.Expr, env *Env) (reflect.Value, error) {
 				return reflect.ValueOf(toString(lhsV) + toString(rhsV)), nil
 			}
 			if (lhsV.Kind() == reflect.Array || lhsV.Kind() == reflect.Slice) && (rhsV.Kind() != reflect.Array && rhsV.Kind() != reflect.Slice) {
+				rhsT := rhsV.Type().Elem()
+				lhsT := lhsV.Type().Elem()
+				if lhsT.Kind() != rhsT.Kind() {
+					if !rhsT.ConvertibleTo(lhsT) {
+						return NilValue, NewStringError(expr, "Unknown operator")
+					}
+					rhsV = rhsV.Convert(lhsT)
+				}
 				return reflect.Append(lhsV, rhsV), nil
 			}
 			if (lhsV.Kind() == reflect.Array || lhsV.Kind() == reflect.Slice) && (rhsV.Kind() == reflect.Array || rhsV.Kind() == reflect.Slice) {
+				rhsT := rhsV.Type().Elem()
+				lhsT := lhsV.Type().Elem()
+				if lhsT.Kind() != rhsT.Kind() {
+					if !rhsT.ConvertibleTo(lhsT) {
+						return NilValue, NewStringError(expr, "Unknown operator")
+					}
+					for i := 0; i < rhsV.Len(); i++ {
+						lhsV = reflect.Append(lhsV, rhsV.Index(i).Convert(lhsT))
+					}
+					return lhsV, nil
+				}
 				return reflect.AppendSlice(lhsV, rhsV), nil
 			}
 			if lhsV.Kind() == reflect.Float64 || rhsV.Kind() == reflect.Float64 {

--- a/vm/vm_test.go
+++ b/vm/vm_test.go
@@ -440,3 +440,37 @@ func TestTypedNil(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestSliceConvert(t *testing.T) {
+	env := NewEnv()
+
+	input := []string{"one", "two"}
+	err := env.Define("input", input)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = env.Execute(`
+	output = []
+	output += input
+	output += [3]
+	`)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	value, err := env.Get("output")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if value.Index(0).Interface().(string) != "one" {
+		t.Fatal("want string `one` but got %v", value.Index(0).Interface())
+	}
+	if value.Index(1).Interface().(string) != "two" {
+		t.Fatal("want string `two` but got %v", value.Index(1).Interface())
+	}
+	if value.Index(2).Interface().(int64) != 3 {
+		t.Fatal("want int 3 but got %v", value.Index(2).Interface())
+	}
+}

